### ISSUE TITLE
Snoop 2.9 crashes on startup

### ISSUE THREAD

**Issue by tiesmaster** [cplotts/snoopwpf#88]
Ey, Snoop stays the most awesome WPF inspection tool. It's not working for me atm, and I missed it the entire day ;)

However, it suddenly stopped working :(

# Problem
When I start Snoop, it doesn't show up. After a bit of digging, I found out that so-called AppCrash reports are being created (here: `C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_snoop.exe_37e7083a3a2cc8e51f6b2f47580bbdb1cf5277_00000000_3cb3fea2`), and in the Event Log I also found entries about the crashes (see below).

# Own steps to fix things
I wasn't sure if my install was faulty, or maybe something wrong with my installed version, so I uninstalled the one that was installed, and re-installed the latest version, and it still didn't show up.

# Environment
I'm running Windows 10, upgraded to the latest edition. Visual Studio 2017 (latest one 15.8.*), full .NET Framework _at least_ 4.6.2.

# Windows Event Entry
````
Fault bucket 1364462021916346858, type 5
Event Name: CLR20r3
Response: Not available
Cab Id: 0

Problem signature:
P1: snoop.exe
P2: 2.9.0.0
P3: 5b5b9b0b
P4: PresentationCore
P5: 3.0.0.0
P6: 5a7d2bb4
P7: 21db
P8: a
P9: System.NullReferenceException
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERF281.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERF31E.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERF32F.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERF35F.tmp.txt

These files may be available here:
C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_snoop.exe_a0bc4fb2736589139e7b5322443942330c8a_00000000_68c6f919

Analysis symbol: 
Rechecking for solution: 0
Report Id: 536d7e93-f7ea-4185-87da-2a9f2ea2164c
Report Status: 268435456
Hashed bucket: 5675642891bd535aa2ef8afacdd5c9ea
Cab Guid: 0
````

**Comment by batzen** [comment]
There should be two eventlog entries. One in System and one in Software.
The one you attached does only contain exception name.
If there is nothing you could try to start snoop with a debugger attached and try to extract some information that way.

**Comment by tiesmaster** [comment]
@batzen Thnx for the swift reply!

I just started to continue the investigation, and wanted fresh entries in the event viewer, for easy searching, and now snoop is showing up straight away 😌

I guess reinstalling, **and** turning off Windows and starting it again after the weekend did its trick. Again, thnx for the help.

**Comment by cplotts** [comment]
Glad it finally worked for you @tiesmaster. Some people were reporting issues with Windows 10 blocking Snoop's files (https://github.com/cplotts/snoopwpf/issues/76) ... I wonder if it was that. However, restarting your machine shouldn't have fixed that issue ... I believe most people had to manually go unblock the files.

**Comment by tiesmaster** [comment]
Indeed, that cannot be the case ;)

I thought the rebooting fixed the issue (for good), but it comes back (now and again). I just tried starting Snoop with the debugger, and that gave an interesting stack trace:

````
>	PresentationCore.dll!System.Windows.Interop.HwndTarget.CriticalHandleWMGetobject(System.IntPtr wparam, System.IntPtr lparam, System.Windows.Media.Visual root, System.IntPtr handle)	Unknown	Non-user code. Skipped loading symbols.
 	PresentationCore.dll!System.Windows.Interop.HwndTarget.HandleMessage(int msg, System.IntPtr wparam, System.IntPtr lparam)	Unknown	Non-user code. Skipped loading symbols.
        [30+ frames]
 	snoop.exe!Snoop.SnoopWindowUtils.LoadWindowPlacement(System.Windows.Window window, Snoop.WINDOWPLACEMENT? windowPlacement)	Unknown	Non-user code. Skipped loading symbols.
 	snoop.exe!Snoop.AppChooser.OnSourceInitialized(System.EventArgs e)	Unknown	Non-user code. Skipped loading symbols.
        [20+ frames]
 	PresentationFramework.dll!System.Windows.Application.RunInternal(System.Windows.Window window)	Unknown	Non-user code. Skipped loading symbols.
 	snoop.exe!Snoop.App.Main()	Unknown	Non-user code. Skipped loading symbols.
````

I can't run the project from Visual Studio, since I'm missing the Windows 8.1 SDK, though the stack trace revealed to me that placing the Snoop toolbar at the previous location failed (and I have a 3 monitor setup, including a laptop display. So it must be related to storing that. I'm now going to find out if I can locate the place where Snoop stores it's placement ;)

**Comment by tiesmaster** [comment]
Alright, I checked the [building instructions](https://github.com/cplotts/snoopwpf#how-do-i-build-snoop), and saw it should build out of the box, so spend some time getting it working on my machine (latest Windows 10 machine), and I needed to do [this](https://github.com/cplotts/snoopwpf/compare/master...tiesmaster:88/investigate?expand=1). I'm guessing you're running Windows 8?

Anyhow, that builds the solution, and actually runs snoop without the exception, so bad news is: I cannot investigate this further, good news: I can continue debugging my weird issue in our legacy WPF app 😂

**Comment by cplotts** [comment]
https://github.com/cplotts/snoopwpf/compare/master...tiesmaster:88/investigate?expand=1

Ooooh. Interesting. Maybe we missed something here. I'll look into this.

Your bad news sounds like good news for you. lol. :)

**Comment by jogibear9988** [comment]
maybe you can debug using dnspy

**Comment by tiesmaster** [comment]
@jogibear9988 Sure. What is dnspy?

**Comment by alhalama** [comment]
dnSpy is a debugger and assembly editor.  There are more details on [dnSpy on Github](https://github.com/0xd4d/dnSpy).

**Comment by alhalama** [comment]
@tiesmaster The CriticalHandleWMGetobject has to do with UI automation and it is would seem that this is happening anytime the window position is set whether it is set by Snoop or not.  It might be helpful to see the full call stack including the missing frames.  

Is there anything different that you noticed about when it worked versus when it didn't?  By any chance did you use a screen reader, touch screen, or something else that would have used UI Automation when it failed?

**Comment by tiesmaster** [comment]
@alhalama Thnx. dnSpy looks really awesome, that's exactly the tool I could have used to investigate this problem. However, snoop is not crashing anymore, at all (so far). I don't have the full stack anymore, so it's a dead end, from that perspective...

As for the background, my laptop has a touch screen, and although I usually don't have the Snoop toolbar on the touch screen monitor, I might have had it the last time I used it...

Funny thing, is that I stumbled upon dnSpy a while ago, so the name already looked familiar. I know I was think: this is really awesome, can be useful in certain situations ;)

**Comment by tiesmaster** [comment]
@alhalama It started crashing for me again (after switching my laptop to use a different screen, a big TV for a demo). I'm debugging the app now, and I'm in `SnoopWindowUtils.LoadWindowPlacement()`, with the following values. Do you see something weird here?

![image](https://user-images.githubusercontent.com/460904/48076107-d230b280-e1e4-11e8-8629-bab01c2bb540.png)

**Comment by batzen** [comment]
LoadWindowPlacement should not be able to crash the app as it's wrapped in a try/catch.
What happens if you hit F10 again?

**Comment by tiesmaster** [comment]
Sorry, forgot all about your question. Pressing F10 here, will throw the exception. Indeed, it's really weird that the app crashes, but thinking about it, I think it _ends_, not crashes, because I don't get a Windows screen like "this app crashes", Snoop just doesn't start up (which is what you'd expect from looking at the code).

Further, the popup from dnSpy about the exception actually talked about "unhandled" exception, so that makes it more curious (but might be related to how dnSpy works, and how it presents stuff to the user, I don't have any experience with dnSpy, obviously). That leads to the thought: this is not a normal CLR exception, but Win32, so that might also influence things...

TL;DR Many mysterious things for me, however, Snoop is starting up again, so can't investigate now 😕

**Comment by tiesmaster** [comment]
@alhalama Alright, it's crashing again 🎉 😉 and I ran procdump to capture the crash, I've uploaded it [here](https://drive.google.com/file/d/1WNJQCeY9WwaGobb2xsThTvDC2FQdz5h8/view?usp=sharing).

Happy Hackin'

**Comment by tiesmaster** [comment]
BTW I'm curious to your results (and how you're investigating this ;)). Keep me posted please.

**Comment by batzen** [comment]
Could you try to reproduce your issue with snoop 2.10 or the current source build from the master branch? Your memory dump doesn't even contain snoop in the crash path...

**Comment by tiesmaster** [comment]
Ah, there is a newer version out, didn't noticed that ;) I've installed that one, and it's still breaking there, but in a weirder way 😕 (the main toolbar pops-up, but completely black, and disappears really quickly).

However, it does show up when I build from latest source 🎉 So at this stage, I think it's more related to some weird stuff happening on my machine, and since I'm going to switch jobs in a couple weeks anyhow, and hence, also machines, and also no WPF more, I'm going to close down this issue, to save everyone some time ;)
